lego_manager/lego_manager.py

```diff
@@ -42,94 +53,122 @@ def on_disconnect(self, conn: rpyc.Connection) -> None:
             'elephant': 'Lego3.example.components.elephant.Elephant'
         }
-        hostnames = (hostname.strip() for hostname in query.split('and'))
+        components = (component.strip() for component in query.split('and'))
+
+        return {component: components_to_class_path[component.split('.')[0]] for component in components}
+
+    @staticmethod
+    def _run_query(query: str) -> str:
+        """Find available setup according to given query.
+
+        Runs the requested setup query on the setup management files, and add instance names to components
+        with unspecified names.
+        For example, query given - 'zebra.alice and elephant', query returned - 'zebra.alice and elephant.bob'.
 
-        return [(hostname, hostname_to_lib[hostname]) for hostname in hostnames]
+        Args:
+            query: A query that describes the desired setup.
+
+        Returns:
+            Final components query for the test. Each component should be built from component class and instance name.
+        """
+        return query
 
-    def exposed_acquire(self, query: str, exclusive: bool): # type: ignore
+    def exposed_acquire_setup(self, query: str, exclusive: bool):  # type: ignore
         """Acquired the desired setup if available.
 
         This function also will store all of the data about the setup usage.
 
         Args:
-            query: A query describes the desired setup.
-            exclusive: Weather the required setup needed exclusivly.
+            query: A query that describes the desired setup.
+            The query syntax is -
+                1. The components should be splitted by the word 'and'.
+                2. The format should be <component_class>.<instance_name>, or only <component_class> if
+                specific instance isn't needed.
+            Example for a query - 'zebra.alice and elephant.bob and giraffe.4'.
+
+            exclusive: Whether the required setup is needed exclusively.
 
         Returns:
-            Allocated requested setup as list of hostnames and libraries.
+            Allocated requested setup as list of tuples made of
+            components names and corresponding paths to Components classes.
+            e.g. [('zebra.alice', 'Lego3.example.components.zebra.Zebra'), ...]
         """
-
         return self._allocation(LegoManager._run_query(query), exclusive)
 
 
 def main() -> None:
     """Starts Lego server."""
 
     rpyc.lib.setup_logger()
-    from rpyc.utils.server import ThreadedServer # pylint: disable=import-outside-toplevel
+    from rpyc.utils.server import ThreadedServer  # pylint: disable=import-outside-toplevel
     # Note: all connection will use the same LegoManager
-    lego_server = ThreadedServer(LegoManager(), port=18861)
+    lego_server = ThreadedServer(
+        LegoManager(),
+        port=LegoManager.DEFAULT_PORT,
+        protocol_config={'allow_public_attrs': True}
+    )
```

Comment:
remove the allow_public_attrs
```suggestion
        port=LegoManager.DEFAULT_PORT
    )
```

Reply:
exposed_acquire_setup (the only method the service exposes) returns a dictionary. To allow use of public attributes of this dict in components_factory, we have to define 'allow_public_attr': True.
It's allows calling .items() on the returned dict (which is actually a netref object).